File: study_ws/src/cpp_pkg/src/NumberCounter.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"
#include "std_srvs/srv/set_bool.hpp" 

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounter : public rclcpp::Node {
public:
    NumberCounter() : Node("number_counter"), count_(0), history_(0) {
        subscriber_ = this->create_subscription<std_msgs::msg::Int64>(
            "number", 10, std::bind(&NumberCounter::callbackNumber, this, std::placeholders::_1));

        publisher_ = this->create_publisher<std_msgs::msg::Int64>("number_count", 10);

        counter_reset_service_ = this->create_service<std_srvs::srv::SetBool>(
            "reset_counter", std::bind(&NumberCounter::reset_callback, this, _1, _2));
    }

private:
    void callbackNumber(const std_msgs::msg::Int64::SharedPtr msg) {
        if(msg->data != history_){
            count_ += 1;
            auto new_msg = std_msgs::msg::Int64();
            new_msg.data = count_;
            publisher_->publish(new_msg);
            history_ = msg->data;
            RCLCPP_INFO(this->get_logger(), "Received: '%ld', Instance: '%ld'", msg->data, count_);
        }
    }

    void reset_callback(const std::shared_ptr<std_srvs::srv::SetBool::Request> request,
                        std::shared_ptr<std_srvs::srv::SetBool::Response> response) {   
        if(request->data){
            count_ = 0;
            response->success = true;
            response->message = "counter_resetted...";
            RCLCPP_INFO(this->get_logger(),"%s", response->message.c_str());
        }
        else{
            response->success = false;
            response->message = "counter_not_resetted...";
        }
    }

    rclcpp::Subscription<std_msgs::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<std_srvs::srv::SetBool>::SharedPtr counter_reset_service_;
    int64_t count_;
    int64_t history_;
};

int main(int argc, char *argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<NumberCounter>());
    rclcpp::shutdown();
    return 0;
}

File: study_ws/src/py_pkg/py_pkg/numberCounter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64
from std_srvs.srv import SetBool

class CounterNode(Node): 
    def __init__(self):
        super().__init__("numberCount")
        self.get_logger().info('Number Counter Has Initiated...')
        self.counter, self.history = 0, 0
        self.counter_reset_service = self.create_service(SetBool, 'reset_counter', self.counter_reset)
        self.subscription = self.create_subscription(Int64, "number", self.subscriber_callback, 10)

    def subscriber_callback(self, msg):
        if self.history != msg.data:
            self.counter += 1
            self.history = msg.data
        self.get_logger().info("Data: " + str(msg.data) + " | Instance: " + str(self.counter))

    def counter_reset(self, request, response):
        if request.data:
            self.counter = 0
            response.success = True
            response.message = "Counter_has_been_reset"
        else:
            response.success = False
            response.message = "Counter_has_not_reset"
        return response 

def main(args=None):
    rclpy.init(args=args)
    node = CounterNode() 
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()
